dotcom-rendering/src/model/buildLightboxImages.ts

```diff
@@ -0,0 +1,110 @@
+ *
+ */
+const isLightboxable = (image: EnhancedImageForLightbox): boolean => {
+	const fields = image.media.allImages[0]?.fields;
+	if (!fields) return false; // Unlikely
+	const { width, height } = fields;
+	const orientation =
+		parseInt(width) > parseInt(height) ? 'horizontal' : 'portrait';
+	switch (orientation) {
+		case 'horizontal':
+			// If any width is above 620 we allow this image in lightbox
+			return image.media.allImages.some(
+				(mediaImg) => parseInt(mediaImg.fields.width) > 620,
+			);
+		case 'portrait':
+			// If any height is above 620 we allow this image in lightbox
+			return image.media.allImages.some(
+				(mediaImg) => parseInt(mediaImg.fields.height) > 620,
+			);
+	}
+};
+
+/**
+ * Generates a new array of lightbox images. Does not mutate.
+ *
+ * We decide this array, prior to rendering, because we create and
+ * insert the lightbox html in the dom at page load and to do this
+ * we need to know which images we want the lightbox to contain.
+ *
+ * This enhancer needs to run at a higher level to most other enhancers
+ * because it need both mainMediaElements and blocks in scope. Most other
+ * enhancers just have blocks in scope.
+ */
+export const buildLightboxImages = (
+	format: FEFormat,
+	blocks: Block[],
+	mainMediaElements: FEElement[],
+): EnhancedImageForLightbox[] => {
+	const images: EnhancedImageForLightbox[] = [];
+	mainMediaElements.forEach((element) => {
+		if (
+			element._type ===
+			'model.dotcomrendering.pageElements.ImageBlockElement'
+		) {
+			images.push(element);
+		}
+	});
+	blocks.forEach((block) => {
+		block.elements.forEach((element) => {
+			let imageElement: EnhancedImageForLightbox;
+			switch (element._type) {
+				case 'model.dotcomrendering.pageElements.ImageBlockElement': {
+					imageElement = element;
+					if (
+						format.design === 'LiveBlogDesign' ||
+						format.design === 'DeadBlogDesign'
+					) {
+						imageElement.blockId = block.id;
+						imageElement.firstPublished = block.blockFirstPublished;
+					}
+					images.push(imageElement);
+					break;
+				}
+				case 'model.dotcomrendering.pageElements.MultiImageBlockElement': {
+					element.images.forEach((multiImage) => {
+						imageElement = multiImage;
+						if (
+							format.design === 'LiveBlogDesign' ||
+							format.design === 'DeadBlogDesign'
+						) {
+							imageElement.blockId = block.id;
+							imageElement.firstPublished =
+								block.blockFirstPublished;
+						}
+						images.push(imageElement);
+					});
+					break;
+				}
+				default:
+					break;
+			}
+		});
+	});
+	// On some articles the main media is repeated as an element in the article body so
+	// we deduplicate the array here
+	const alreadySeen: string[] = [];
+	const uniqueImages: EnhancedImageForLightbox[] = [];
+	images.forEach((image) => {
+		const imageId = image.media.allImages[0]?.url.split('/')[3];
```

Comment:
Is there a safer way to check this is a valid ID, since we're relying on 'arbitrary' indexes, perhaps a regex or something that matches what we know IDs to look like?

Reply:
You're not wrong, this code is dependent on a bunch of things. I have a feeling it's like this because Frontend did something similar but I can't find it. The obvious thing for deduplication would have been to check the whole url but I suspect this won't work - hence why I'm digging into the url to get the id - but I can't remember why I suspect this! Sorry, vague answer, I forgot my motivations here!

I'm not against a regex but I guess that it would have to be checking `image.media.allImages[0]?.url`? And at that point we're already depending on the first image existing and having the expected properties so maybe it doesn't add much but I'm rubbish at regexs so there might be a neater solution

Reply:
[JavaScript `Set`s](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) are well suited to reduce duplication, and the [URL object](https://developer.mozilla.org/en-US/docs/Web/API/URL) can easily strip out any query parameters using the [`pathname` property](https://developer.mozilla.org/en-US/docs/Web/API/URL/pathname).